### DISCUSSION TITLE
How to set custom request headers on trpc client Programmatically ?

### DISCUSSION THREAD

**Discussion by soulsam480** [trpc/trpc#1408] [Q&A] [answered]
I've been looking into this for over 2 hrs now. Is it possible to set custom headers after a query is successfull ?

```ts
// mutation
  .mutation('login', {
    input: z.object({
      email: z.string().email(),
      password: z.string(),
    }),
    async resolve({ ctx, input: { email, password } }) {
      const account = await ctx.prisma.account.findFirst({ where: { email } });

      if (!account)
        throw new TRPCError({
          code: 'NOT_FOUND',
          message: 'Account doeesn"t exist with the provided email',
        });

      const isSamePassword = await comparePassword(password, account.password);

      if (!isSamePassword)
        throw new TRPCError({
          code: 'BAD_REQUEST',
          message: 'Invalid email or password',
        });

      const { accessToken, refreshToken } = createTokens(`${account.id}`);

      ctx.res.setHeader('Set-Cookie', refreshToken);

      return {
        ...account,
        accessToken,
      };
    },
  });
```
```ts
// mutation client
  const loginMut = trpc.useMutation(['auth.login'], {
    onSuccess() {
      // set header here
    },
  });
```
as you can see I'm setting the refresh-token cookie and sending the accessToken. I want to set the accessToken header on the client after the login mutation is successful. I've looked into the docs of trpc and react-query but failed to find the solution. I'm new to trpc but love the concept and DX. Can anybody help ?

**Comment by timcole** [comment, marked as answer]
The `headers` option in config when using [`withTRPC`](https://trpc.io/docs/nextjs#config-callback) with next or [`createClient`](https://trpc.io/docs/react#4-add-trpc-providers) with normal react might be useful here.

`headers` can be both an object or a function. If it’s a function it gets called every request. So you could set it to something like.

```ts
export let token: string;
```

```ts
export default withTRPC<AppRouter>({
  config({ ctx }) {
    headers() {
      return {
        Authorization: token
      };
    },
  }
})(MyApp);
```

```ts
const loginMut = trpc.useMutation([‘auth.login’], {
  onSuccess({ accessToken }) {
    token = accessToken;
  },
});
```

Where `token` can be whatever you want to get the token. Whether that’s just a client-side variable that you update the value of on success like my example or you store the token and pull it from something like localStorage, completely up to you how you wanna do that.

**Reply by soulsam480** [reply]
Hey thanks for the really quick answer, so the token variable will be reactive right ? I mean changing the value will also change the reference inside withTrpc function or not ?

**Reply by KATT** [reply]
Not sure what you mean by reactive, but the `headers()`-callback is called on every http call so it's dynamic

**Reply by soulsam480** [reply]
Ahh, thanks for this. Will do it. Grateful for the project. Thanks

**Comment by piotrski** [comment]
@KATT BTW what is the reasoning behind `withTRPC` not passing all the request headers in ssr mode automatically?

**Reply by KATT** [reply]
Comment here: #1811 :)
